### ISSUE TITLE
mkdir in profile does not respect --private=<directory>

### ISSUE THREAD

**Issue by pyamsoft** [netblue30/firejail#903]
Latest stable firejail: 0.9.44

The command line option `--private=<directory>` states that firejail uses the given directory as a new home. 
```
--private=directory
      Use directory as user home.

      Example:
      $ firejail --private=/home/netblue/firefox-home firefox
```

This works fine, and all whitelist and blacklist commands work as expected with respect to the private directory. However, using mkdir in a profile incorrectly creates the directory in the real home.

You can observe this in the shipped Dropbox profile for example, which makes calls to mkdir in its profile:
```
firejail --private=/home/pyamsoft/dropboxtesting dropbox
```

You can observe similar errors in the shipped firefox profile. 
```
firejail --private=/home/pyamsoft/firefoxtesting firefox
```

You will see that a Dropbox folder will be created in the actual home directory, instead of the private home at ~/dropboxtesting. You will also see the firefox directories like ~/.mozilla being created in the real home. I would expect that mkdir will respect the private home directory just as whitelist and blacklist do, so I believe this behavior to be a bug.

**Comment by valoq** [comment]
I can confirm the behaviour. Seems like a bug to me.

**Comment by netblue30** [comment]
Fixed, thanks for the bug.

**Comment by okcomputerik** [comment]
Although marked as fixed, I see the same behavior running firejail 0.9.44.8 with the dropbox profile. I'm I missing something here? Thanks!

**Comment by netblue30** [comment]
I'll look into it, thanks.

**Comment by sunho0301** [comment]
Are there any updates on this issue?? I'm having the same problem.
Thanks!

**Comment by chiraag-nataraj** [comment]
Is this still an issue? If so, let's figure out what's going on :)

**Comment by chiraag-nataraj** [comment]
Yes, this is still an issue. The question becomes the following:
If we write `mkdir ${HOME}/<somedir>`, do we mean the _real_ `${HOME}` or what the _sandbox_ will see as `${HOME}`?

**Comment by chiraag-nataraj** [comment]
@netblue30 Is this intended behavior? If we use `${HOME}` in, say, `mkdir`, do we want `${HOME}` to be resolved _after_ things like `--private` have been resolved or do we want to resolve it _before_ those predicates? That will change the behavior as I understand it, and either one is fine as long as we document it.

**Comment by chiraag-nataraj** [comment]
Sorry to dredge up so many old issues, but @netblue30, is this intended behavior? If so, we can close. Otherwise, we should fix this.

**Comment by matu3ba** [comment]
@rusty-snake @chiraag-nataraj 
If the folder does not exist, firejail prompts `Error: invalid private directory` and exists. If nesting of firejails is [forbidden](https://github.com/netblue30/firejail/issues/3333#issuecomment-611510025), the error could maybe be simplified to `Error: folder not usable` or something like this.  
Otherwise suggestion to close this.

**Comment by rusty-snake** [comment]
@matu3ba it's not about the private directory does not work. The mkdir does not respect it. Like this:
```
$ cat tmp.profile
mkdir ${HOME}/foo
private
mkdir ${HOME}/bar
$ firejail --profile=tmp.profile true
$ [ -e foo ] && echo "foo exists"
foo exists
$ [ -e bar ] && echo "bar exists"
bar exists
```

**Comment by colons** [comment]
hi. i don't know that this issue has any actual security consequences, but it is _extremely_ disconcerting to see your home directory littered with directories related to something you launched with the promise that it'd be contained to a directory you specified

**Comment by spikethehobbitmage** [comment]
> Maybe we could postpone `mkdir` and `mkfile`until after private option processing, but handle them before `whitelist`.

This is exactly what needs to happen. `overlay` options should* also be processed before `mk*` since they also discard jailed changes.

*This may already be the case but `overlay` is disabled on my system so I can't test it.

**Comment by giddie** [comment]
This is disconcerting, and I'm trying to understand the scope of the issue. If I do this:

```bash
$ mkdir ~/jail
$ firejail --private=~/jail bash -c "mkdir ~/another"
$ ls ~/jail
another
$ ls ~/another
ls: cannot access '/home/user/another`: No such file or directory
```

So it actually seems to be working correctly. How do we reproduce the bug?

**Comment by tanriol** [comment]
> How do we reproduce the bug?

`firejail --private=~/jail --mkdir=~/another bash -c "true"`

This looks kinda stupid when both are specified on the command line - why do we need `mkdir` in a private profile - but when one is `mkdir` in `steam.profile` (dozens of them are there) and the other is `--private=${HOME}/steam-jail` in the script used to launch Steam, the fact that it still creates dirs in real home like `${HOME}/Zomboid` (not even a hidden one!) feels plain wrong.

**Comment by giddie** [comment]
Aaaaah! 💡 This is a _lot_ less concerning than the manpage suggests. When it says "some *commands* such as mkdir ...", what I understood was that `mkdir` _inside_ the container was broken. OK, this is not such a huge deal. But the bug warning could maybe better communicate that it's just referring to firejail command-line arguments or commands in _profiles_.
